compiler/src/dotty/tools/dotc/transform/init/Semantic.scala

```diff
@@ -600,6 +600,24 @@ object Semantic {
         || (meth eq defn.Object_ne)
         || (meth eq defn.Any_isInstanceOf)
 
+      def checkArgsWithParametricity() =
+        val methodType = atPhaseBeforeTransforms { meth.info.stripPoly }
+        var allArgsPromote = true
+        var allParamTypes = methodType.paramInfoss.flatten.map(_.repeatedToSingle)
+        val errors = allParamTypes.zip(args).flatMap { (info, arg) =>
+          val isTypeParam = info.isInstanceOf[TypeParamRef]
+          val isWithinBounds = info.bounds.lo <:< defn.NothingType && defn.AnyType <:< info.bounds.hi
+          def otherParamContains = allParamTypes.exists( param => { param != info && param.typeSymbol != defn.ClassTagClass && info.occursIn(param) } )
```

Comment:
Nitpick:
```suggestion
          def otherParamContains = allParamTypes.exists { param => param != info && param.typeSymbol != defn.ClassTagClass && info.occursIn(param) } 
```

Reply:
BTW, there seems to be a typo: `param.occursIn(info)` instead of `info.occursIn(param)`?

What about adding the following as a neg test?

```Scala
object A:
  def foo[T, S <: T](x: S, array: Array[T]): Unit = array(0) = x

class B:
  var a = new Array[B](2)
  A.foo(this, a) // error
  println(a(0).i)
  val i = 99
```

Reply:
> BTW, there seems to be a typo: `param.occursIn(info)` instead of `info.occursIn(param)`?

`info.occursIn(param)` is correct as we want to check whether the type `T` occurs in any of the other parameters.

> What about adding the following as a neg test?

That test was already added in https://github.com/lampepfl/dotty/pull/14895. It passes with the current code

Reply:
@liufengyun Just tried adding the test and it fails due to not triggering a warning. This test case is more difficult to handle because we can't rely on the `occursIn` function.

After this PR is merged, I can open an issue to handle this test case.

Reply:
> we can't rely on the occursIn function

That is a little strange. BTW, for the new example, `isWithinBounds` should be false(the upper bound of `S` is `T`, which is not `Any`), thus no reasoning by parametricity will be used here.

It seems the problem lies in how to retrieve bounds of a `TypeParamRef`. `info.bounds` will not return the bounds, the following will help:

```Scala
info match
case typeParamRef: TypeParamRef =>
    val bounds = typeParamRef.info.bounds
case _ =>
 ```